README
++++++++++++++
D-BUS Exercise
++++++++++++++

:Author: Gao Nan

Introduction
============

This project contains three applications including one receiver and two senders. 
The applications designed using low-level D-Bus C API. So they could communicate 
with each other over DBUS. 

One sender will send a signal every one second to inform the receiver to randomly 
add an item to the queue that holds items (1024 default) that consists of a name 
and a value. The other sender will invoke a remote method every five seconds to 
remove an item from the queue randomly. 

The applications run until 'CTRL-C' is hit. Especially, the receiver will exit 
if the queue is full or empty.

Installation
============

To install this project, you should first unzip the tarball to a directory, and 
enter the directory to type the command "./configure && make && sudo make install". 
The applications will be compiled and installed to your system directory 
("/usr/local/bin" default) which you can change when running the command "./configure" 
using the configure option '--prefix='.

Then you could type commands "receiver", "sender_add" or "sender_remove" to run 
the collection of applications which displays what it is doing.

If there is a warning says 
"Connection Error (Failed to execute dbus-launch to autolaunch D-Bus session)" 
when you want to run applications, do the following to solve it: 
* type command "dbus-launch" 
* export the given paths 


NEWS
D-Bus Exercise 1.0
===

 - Initial release.


src/App2/sender_add.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dbus/dbus.h>
#include "queue.h"

/*
 * Connect to the DBUS bus and send a signal 
 * to add items to the queue at a random position
 */
void send_signal_add_items(Item *item)
{
	DBusMessage *msg;
	DBusConnection *conn;
	DBusError err;
	int ret;
	dbus_uint32_t serial = 0;

	// initialize the error value
	dbus_error_init(&err);

	// connect to the DBUS system bus, and check for errors
	conn = dbus_bus_get(DBUS_BUS_SESSION, &err);
	if (dbus_error_is_set(&err))
	{ 
		fprintf(stderr, "Connection Error (%s)\n", err.message); 
		dbus_error_free(&err); 
	}
	if (conn == NULL)
	{ 
		exit(1);
	}

	// register our name on the bus, and check for errors
	ret = dbus_bus_request_name(conn, "queue.msg.add", DBUS_NAME_FLAG_REPLACE_EXISTING, &err);
	if (dbus_error_is_set(&err))
	{ 
		fprintf(stderr, "Name Error (%s)\n", err.message); 
		dbus_error_free(&err); 
	}
	if (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER != ret)
	{ 
		exit(1);
	}

	// loop listening for signals being emmitted
	while (1)
	{
		// create a signal & check for errors 
		msg = dbus_message_new_signal("/queue/msg/Object",  // object name of the signal
		                              "queue.msg.Handler",  // interface name of the signal
					      "Add");               // name of the signal
		if (msg == NULL) 
		{ 
			fprintf(stderr, "Message Null\n"); 
			exit(1); 
		}

		// append arguments onto signal
		char *s = (char *)malloc(512);
		strcpy(s, item->name);
		dbus_uint32_t v = item->value;
		dbus_message_append_args(msg, 
		                         DBUS_TYPE_STRING, &s,  // the name of the item
					 DBUS_TYPE_UINT32, &v,  // the value of the item
					 DBUS_TYPE_INVALID);
		free(s);
		s = NULL;

		// send the message and flush the connection
		if (!dbus_connection_send(conn, msg, &serial))
		{
			fprintf(stderr, "Out Of Memory!\n"); 
			exit(1);
		}
		dbus_connection_flush(conn);

		printf("Signal sent to add an item with name \"%s\" valued (%d)\n", item->name, item->value);

		// free the message 
		dbus_message_unref(msg);

		// wait for one second
		sleep(1);
	}
}

int main(int argc, char *argv[])
{
	// initialize an item that consists of a name and a value, 
	// and send it to the DBUS
	Item *item = (Item *)malloc(sizeof(Item));
	strcpy(item->name, "add");
	item->value = 1;
	send_signal_add_items(item);
	free(item);
	item = NULL;
	return 0;
}


src/App3/sender_remove.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dbus/dbus.h>
#include "queue.h"

/*
 * Connect to the DBUS bus and send a signal 
 * to remove items from the queue randomly
 */
void send_signal_remove_items()
{
	DBusMessage *msg;
	DBusConnection *conn;
	DBusError err;
	int ret;
	dbus_uint32_t serial = 0;

	// initialize the error value
	dbus_error_init(&err);

	// connect to the DBUS system bus, and check for errors
	conn = dbus_bus_get(DBUS_BUS_SESSION, &err);
	if (dbus_error_is_set(&err))
	{ 
		fprintf(stderr, "Connection Error (%s)\n", err.message); 
		dbus_error_free(&err); 
	}
	if (conn == NULL)
	{ 
		exit(1);
	}

	// register our name on the bus, and check for errors
	ret = dbus_bus_request_name(conn, "queue.msg.remove", DBUS_NAME_FLAG_REPLACE_EXISTING, &err);
	if (dbus_error_is_set(&err))
	{ 
		fprintf(stderr, "Name Error (%s)\n", err.message); 
		dbus_error_free(&err); 
	}
	if (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER != ret)
	{ 
		exit(1);
	}

	// loop listening for signals being emmitted
	while (1)
	{
		// create a method cal & check for errors 
		msg = dbus_message_new_method_call("queue.msg.receiver",  // destination name
		                                   "/queue/msg/Object",   // object name of the signal
		                                   "queue.msg.Handler",   // interface name of the signal
						   "Remove");                // name of the method
		if (msg == NULL) 
		{ 
			fprintf(stderr, "Message Null\n"); 
			exit(1); 
		}

		// send the message and flush the connection
		if (!dbus_connection_send(conn, msg, &serial))
		{
			fprintf(stderr, "Out Of Memory!\n"); 
			exit(1);
		}
		dbus_connection_flush(conn);

		printf("Signal sent to remove an item randomly\n");

		// free the message 
		dbus_message_unref(msg);

		// wait for five seconds
		sleep(5);
	}
}

int main(int argc, char *argv[])
{
	send_signal_remove_items();
	return 0;
}


src/App1/receiver.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <dbus/dbus.h>
#include "queue.h"

#define MAXNUM 1024  // the max number of items in the queue

int num = 0;         // current number of items in the queue
Item queue[MAXNUM];  // global queue for the program

/*
 * Randomly adds items to the queue, 
 * and returns the position if success
 */
int random_add_items(Item queue[MAXNUM], Item *item)
{
	if (num == 0)
	{
		strcpy(queue[0].name, item->name);
		queue[0].value = item->value;
		num = 1;
		return 0;
	}
	if (num >= MAXNUM)
	{
		printf("The queue is full!\n");
		return -1;
	}

	// generate a random number between 0 and current number of the queue
	srand((unsigned int)time(NULL));
	int n = rand() % num;
	int i;
	for (i = num - 1; i >= n; --i)
	{
		strcpy(queue[i + 1].name, queue[i].name);
		queue[i + 1].value = queue[i].value;
	}
	strcpy(queue[n].name, item->name);
	queue[n].value = item->value;
	++num;
	return n;
}

/*
 * Randomly removes items from the queue, 
 * and returns the position if success
 */
int random_remove_items(Item queue[MAXNUM])
{
	if (num <= 0)
	{
		printf("The queue is empty!\n");
		return -1;
	}

	// generate a random number between 0 and current number of the queue
	srand((unsigned int)time(NULL));
	int n = rand() % num;
	int i;
	for (i = n; i < num - 1; ++i)
	{
		strcpy(queue[i].name, queue[i + 1].name);
		queue[i].value = queue[i].value;
	}
	--num;
	return n;
}

/*
 * Receive the signals from the bus
 */
void receive()
{
	DBusMessage *msg;
	DBusMessageIter iter;
	DBusConnection *conn;
	DBusError err;
	int ret;

	Item *item = (Item *)malloc(sizeof(Item));
	if (item == NULL)
	{
		exit(1);
	}
	char *s = (char *)malloc(512 * sizeof(char));
	if (s == NULL)
	{
		exit(1);
	}

	printf("Listening for signals\n");

	// initialize the errors
	dbus_error_init(&err);

	// connect to the bus and check for errors
	conn = dbus_bus_get(DBUS_BUS_SESSION, &err);
	if (dbus_error_is_set(&err))
	{ 
		fprintf(stderr, "Connection Error (%s)\n", err.message);
		dbus_error_free(&err); 
	}
	if (conn == NULL)
	{ 
		exit(1);
	}

	// request our name on the bus and check for errors
	ret = dbus_bus_request_name(conn, "queue.msg.receiver", DBUS_NAME_FLAG_REPLACE_EXISTING , &err);
	if (dbus_error_is_set(&err))
	{ 
		fprintf(stderr, "Name Error (%s)\n", err.message);
		dbus_error_free(&err); 
	}
	if (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER != ret)
	{
		exit(1);
	}

	// add a rule for which messages we want to see
	// see signals from the given interface
	dbus_bus_add_match(conn, "interface='queue.msg.Handler'", &err);
	dbus_connection_flush(conn);
	if (dbus_error_is_set(&err))
	{ 
		fprintf(stderr, "Match Error (%s)\n", err.message);
		exit(1); 
	}

	// loop listening for signals being emmitted
	while (1)
	{
		// non blocking read of the next available message
		dbus_connection_read_write(conn, 0);
		msg = dbus_connection_pop_message(conn);

		// loop again if we haven't read a message
		if (msg == NULL)
		{
			sleep(1);
			continue;
		}

		// signals for adding
		// check if the message is a signal from the correct interface and with the correct name
		if (dbus_message_is_signal(msg, "queue.msg.Handler", "Add"))
		{

			// read the parameters
			if (!dbus_message_iter_init(msg, &iter))
			{
				fprintf(stderr, "Message Has No Parameters\n");
			}
			else
			{
				dbus_uint32_t v;
				dbus_message_iter_get_basic(&iter, &s);
				strcpy(item->name, s);
				dbus_message_iter_next(&iter);  // moves the iter to read next param
				dbus_message_iter_get_basic(&iter, &v);
				item->value = v;
			}

			int n = random_add_items(queue, item);
			if (n != -1)
			{
				printf("Added an item with name \"%s\" valued (%d) at position (%d)\n", 
				       item->name, item->value, n);
			}
			else
			{
				exit(1);
			}
		}

		// methods for removing
		// check if the message is a method call from the correct interface and with the correct name
		else if (dbus_message_is_method_call(msg, "queue.msg.Handler", "Remove"))
		{
			int n = random_remove_items(queue);
			if (n != -1)
			{
				printf("Removed an item from position (%d)\n", n);
			}
			else
			{
				exit(1);
			}
		}

		// free the message
		dbus_message_unref(msg);
	}

	// free the item
	free(item);
	item = NULL;
	// free the s
	free(s);
	s = NULL;
}

int main(int argc, char *argv[])
{
	receive();
	return 0;
}
